From c0197f7c74ac94a5707db598d7445ad68a1bf31b Mon Sep 17 00:00:00 2001
From: Zdenek Kabelac <zkabelac@redhat.com>
Date: Mon, 15 Dec 2025 11:44:40 +0100
Subject: [PATCH 28/32] filesystem: add internal xfs struct copy

When the xfs.h header is missing (as these XFS headers are not commonly
installed), instead of completely disabling this functionality,
let's use our internal copy of the basic structure.
Just to cover our 2 needed variables with the ioctl number.

If this would have ever changed - updated is needed...

Prefered use remains to use xfs/xfs.h.

(cherry picked from commit 0774109830d50c5987bb48c877dd5d2f51e406b5)
---
 lib/device/filesystem_xfs.c | 38 +++++++++++++++++++++++++------------
 1 file changed, 26 insertions(+), 12 deletions(-)

diff --git a/lib/device/filesystem_xfs.c b/lib/device/filesystem_xfs.c
index 61252fe16..a6ab5856d 100644
--- a/lib/device/filesystem_xfs.c
+++ b/lib/device/filesystem_xfs.c
@@ -19,6 +19,31 @@
 
 #include <xfs/xfs.h>
 
+#else /* !HAVE_XFS_XFS_H */
+
+#include <sys/ioctl.h>
+#include <stdio.h>
+#include <unistd.h>
+
+/* #IOCTL to obtain geometry info */
+#define XFS_IOC_FSGEOMETRY 0x8100587e
+
+/* Copy of some basic and hopefully unchangable XFS header data variables */
+struct xfs_fsop_geom {
+	uint32_t blocksize;	/* filesystem (data) block size */
+	uint32_t rtextsize;
+	uint32_t agblocks;
+	uint32_t agcount;
+	uint32_t logblocks;
+	uint32_t sectsize;
+	uint32_t inodesize;
+	uint32_t imaxpct;
+	uint64_t datablocks;	/* fsblocks in data subvolume */
+	uint32_t reserved[128];	/* whatever size whole structure may have */
+};
+
+#endif
+
 int fs_xfs_update_size_mounted(struct cmd_context *cmd, struct logical_volume *lv,
 			       char *lv_path, struct fs_info *fsi)
 {
@@ -40,7 +65,7 @@ int fs_xfs_update_size_mounted(struct cmd_context *cmd, struct logical_volume *l
 	fsi->fs_last_byte = geo.blocksize * geo.datablocks;
 	ret = 1;
 
-	log_debug("xfs geometry blocksize %llu datablocks %llu fs_last_byte %llu from %s %s",
+	log_debug("XFS geometry blocksize %llu datablocks %llu fs_last_byte %llu from %s %s.",
 		  (unsigned long long)geo.blocksize, (unsigned long long)geo.datablocks,
 		  (unsigned long long)fsi->fs_last_byte, lv_path, fsi->mount_dir);
 out:
@@ -48,14 +73,3 @@ out:
 
 	return ret;
 }
-
-#else /* !HAVE_XFS_XFS_H */
-
-int fs_xfs_update_size_mounted(struct cmd_context *cmd, struct logical_volume *lv,
-			       char *lv_path, struct fs_info *fsi)
-{
-	log_debug("No XFS support, continuing WITHOUT reading XFS geometry.");
-	return 1;
-}
-
-#endif
-- 
2.52.0

